### DISCUSSION TITLE
Why zcl_excel_autofilter's method is_row_hidden is not used ?

### DISCUSSION THREAD

**Discussion by darnoc312** [abap2xlsx/abap2xlsx#1229] [Q&A]
Autofilter in Writer's method create_xl_sheet_sheet_data does not work.

First bug:
```abap
        READ TABLE lt_values INTO ls_values WITH KEY column = ls_last_row-cell_column.
        IF sy-subrc = 0 AND ls_values-value = ls_last_row-cell_value.
          CLEAR l_autofilter_hidden.
        ENDIF.
```
This looks always for the first filter value for a certain cell column and compares it to the current cell value.
But this does not work, if there are more than one value per column. They are never considered.
It should look like this:
```abap
        READ TABLE lt_values INTO ls_values WITH KEY column = ls_last_row-cell_column
                                                     value  = ls_last_row-cell_value.
        IF sy-subrc = 0.
          CLEAR l_autofilter_hidden.
        ENDIF.
```

Assuming this bug is fixed there is another bug:
As you see the field l_autofilter_hidden is cleared (means the line is visible) at the first match.
But if there are more than one column filter the line should be visible only if all corresponding cell values match to their respective column filter.

Both bugs could be fixed in the obscure code in method create_xl_sheet_sheet_data.
But i think you like to minimize any change in this coding area.

So why not using the already implemented autofilter's method is_row_hidden, which would does it better ?
And it is more flexible to use because the doing is capsuled in autofilter class itself.
The only thing to change there is the WHILE loop to avoid the creation of empty column filters.
```abap
  METHOD is_row_hidden.

    FIELD-SYMBOLS: <ls_filter> TYPE ts_filter.

    rv_is_hidden = abap_false.

*--------------------------------------------------------------------*
* 1st row of filter area is never hidden, because here the filter
* symbol is being shown
*--------------------------------------------------------------------*
    IF iv_row <= me->filter_area-row_start OR
       iv_row >  me->filter_area-row_end.
      RETURN.
    ENDIF.

*   Therefore the definition of type tt_filters must be changed to SORTED TABLE
    LOOP AT mt_filters ASSIGNING <ls_filter> WHERE column >= me->filter_area-col_start
                                               AND column <= me->filter_area-col_end.

      CASE <ls_filter>-rule.

        WHEN mc_filter_rule_single_values.
          rv_is_hidden = me->is_row_hidden_single_values( iv_row    = iv_row
                                                          iv_col    = <ls_filter>-column
                                                          is_filter = <ls_filter> ).

        WHEN mc_filter_rule_text_pattern.
          rv_is_hidden = me->is_row_hidden_text_pattern(  iv_row    = iv_row
                                                          iv_col    = <ls_filter>-column
                                                          is_filter = <ls_filter> ).

      ENDCASE.

      IF rv_is_hidden = abap_true.
        RETURN.
      ENDIF.

    ENDLOOP.

  ENDMETHOD.
```

So i could propose a PR containing this and the minimum change in method create_xl_sheet_sheet_data (two times), to call this method.
From:
```abap
      IF lo_row->get_visible( ) = abap_false OR
         l_autofilter_hidden = abap_true.
        lo_element_2->set_attribute_ns( name  = 'hidden' value = 'true' ).
      ENDIF.
```
To:
```abap
      IF lo_row->get_visible( ) = abap_false OR
         ( lo_autofilter IS BOUND AND
           lo_autofilter->is_row_hidden( ls_last_row-cell_row ) = abap_true ).
        lo_element_2->set_attribute_ns( name  = 'hidden' value = 'true' ).
      ENDIF.
```

Of course any coding part concerning the field l_autofilter_hidden could be deleted, too. But it can also stay, if you don't like.

**Comment by sandraros** [comment]
I'm really sorry, as I already said, I don't want to discuss FIRST the code because it's messy, bad design, etc., I need FIRST a tiny program which demonstrates the bug, which I can run, reproduce and debug, and then we can discuss. Thank you.

**Comment by darnoc312** [comment]
Sorry if my question wasn't clear.
First: Autofilter in ABAP2XLSX does not work. This is clear and undisputed.
Please refer to #440 and many others. The proposed solution there only works with ZCL_EXCEL_CONVERTER but not without it.
Also, I explained above what the problem is.
If i would propose a change i will open an issue and give you an example, too.
But first of all, i want to know which direction of change you prefer. That's why i opend this discussion.
1) Do you really like to change the obscure code in writer's create_xl_sheet_sheet_data method ? I don't think so.
2) Or do you like to use finally autofilter's own method is_row_hidden ? This method is implemented since years but not used. Are there perhaps any general reservations against it ? I think this is more flexible to use. I would prefer and recommend it.
